### DISCUSSION TITLE
example grammar for a text adventure game

### DISCUSSION THREAD

**Discussion by dcsan** [bikibird/ishml#83] [Q&A] [answered]
Thanks for setting this up!

From my time using ishml, I generally understood the concepts and wanted to use it for a basic parser for a text adventure game. It would need to understand basic sentences such:

```
get (the) $item
open $item
wear $item
drink $item
```
basically
```
$verb (the) $noun
```

then things could get a bit more complex with
`use $item1 on $item2` or adding adjectives.

But I got stuck at actually writing the grammar doc to parse these type of things and just reverted to regex.
As I understand it from here
https://whitewhalestories.com/tutorials/parsing1.html

I would need to specify each possible noun $item separately, is that correct?
or is there some logic to "split" the sentences?
eg if you see

use A on B

then A and B can come back as the result `nouns[]` list.

My comparison for this is how DialogFlow and other NLP tools work
where you define the list of entities (nouns) separately but the system is able to infer similar words.
eg shoe = slipper, sandal etc.

So basically a helpful addition to the library would be a bundled example grammar for basic text adventure type command parsing. 

previous discussion:
https://intfiction.org/t/ishml-new-javascript-library-for-parser-if/41797/33

> What you are missing is a semantics function defined on the nounPhrase to filter out mismatches between adjectives and nouns

seemed really complex setup process compared to a regex with matching.

**Comment by bikibird** [comment, marked as answer]
You're welcome!  **GitHub Discussions** are in beta, but hopefully they are here to stay.

Yes, I'm working on a real parser for text adventures.  Once I get that done it should be more obvious how to connect code to the lexicon.  Feel free to ask about my progress in about a month's time.  You can take a look at [https://github.com/bikibird/ishml/blob/master/extensions/lang.js](https://github.com/bikibird/ishml/blob/master/extensions/lang.js) if you are curious, but this is still very much a work in progress and I don't stand by any of it. 

I got a bit side-tracked for a few months when I decided to add a [text generation system](https://whitewhalestories.com/tutorials/phrases1.html).  Haven't gotten much in the way of [feedback ](https://intfiction.org/t/text-generation-module-added-to-ishml/48606) on this component, but I'm still excited about it.

So, ISHML does require all synonyms to be added to the lexicon manually.  The example you referenced includes multiple synonyms for the "take" key word.  You can register a whole batch of synonyms in one line of code.  I do eventually plan to release a starter kit which will include a parser, standard actions, and other goodies.

With regard to [https://intfiction.org/t/ishml-new-javascript-library-for-parser-if/41797/33](https://intfiction.org/t/ishml-new-javascript-library-for-parser-if/41797/33).  I think I have to take back what I said.  I had made some changes to the code to make things work more simply, but I neglected to update the examples in the tutorial.  Once I realized that, I updated the tutorials, but I think I might not have closed the loop and gotten back to you.  Please accept my apologies.  Anyhow, you should find that the examples are winnowing down the possible interpretations in a more reasonable way now.

**Comment by dcsan** [comment]
yes I think when I tried your suggestion it didn't work for me and I assumed I was doing something wrong but didn't solve it, hence no reply on that thread.

leaving aside the synonyms for individual words - nouns or verbs - which are easy enough to add, is there a basic grammar example for the main sentence structures like the above, and perhaps something a bit more sophisticated?

it's very easy to do a `(get|take) (.*)` regex and obvious what it's doing. no magic or complexity.
If the most simple cases take 100X more time to get working in ishml, it's scary what lies beyond that.
Of course with regex it supposedly doesn't scale, but you can get pretty far with just a set of regex rule => event mapping.

So working code for the simple case, and then some explanation of the benefits over regex I think would help this package get more adoption.

( text generation is a [separate](https://en.wikipedia.org/wiki/Unix_philosophy#Do_One_Thing_and_Do_It_Well) topic )

**Reply by bikibird** [reply]
I don't view text generation as a separate topic.  Yes, separate from parsing, but not separate from a general IF API, which ISHML hopes to be one day.

**Comment by dcsan** [comment]
PS I think using browser examples also hurts adoption, as there's some setup of a web page etc. 
with a server based library you can also write test cases that show how to use the library more easily and add them to the repo. just another $0.02 . that said client debugging is nice.

**Reply by bikibird** [reply]
I used JS Fiddle for the text generation tutorial, but haven't gone back and rewritten the parsing tutorial.
